go-client/client.go

```diff
@@ -0,0 +1,152 @@
+package glpclient
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	DEFAULT_TIMEOUT   = 30
+	DEFAULT_REATTEMPT = 30
+)
+
+type PollResponse struct {
+	Events    []PollEvent `json:"events"`
+	Timestamp int64       `json:"timestamp"`
+}
+
+// Taken from https://github.com/jcuga/golongpoll/blob/master/events.go
+type PollEvent struct {
+	// Timestamp is milliseconds since epoch to match javascrits Date.getTime()
+	Timestamp int64  `json:"timestamp"`
+	Category  string `json:"category"`
+	// NOTE: Data can be anything that is able to passed to json.Marshal()
+	Data json.RawMessage `json:"data"`
+}
+
+type Client struct {
+	url      *url.URL
+	category string
+	// Timeout controls the timeout in all the requests, can be changed after instantiating the client
+	Timeout uint64
+	// Reattempt controls the amount of time the client waits to reconnect to the server after a failure
+	Reattempt time.Duration
+	// Will get all the events data
+	EventsChan chan PollEvent
+	// Flag that tracks the current run ID
+	runID uint64
+	// The HTTP client to perform the requests, any changes on this should be done prior to starting the client the first time
+	HttpClient *http.Client
+	// The username to be used for basic HTTP authentication
+	BasicAuthUsername string
+	// The password to be used for basic HTTP authentication
+	BasicAuthPassword string
+}
+
+// Instantiate a new client to connect to a given URL and send the events into a channel
+// The URL shouldn't contain any GET parameters although its fine if it contains some but category, since_time and timeout will be overriten
+// stubChanData must either be an empty structure of the events data or a map[string]interface{} if the events do not follow a specific structure
+func NewClient(url *url.URL, category string) *Client {
+	return &Client{
+		url:        url,
+		category:   category,
+		Timeout:    DEFAULT_TIMEOUT,
+		Reattempt:  DEFAULT_REATTEMPT * time.Second,
+		EventsChan: make(chan PollEvent),
+		HttpClient: &http.Client{},
+	}
+}
+
+// Start the polling of the events on the URL defined in the client
+// Will send the events in the EventsChan of the client
+func (c *Client) Start() {
+	u := c.url
+	fmt.Println("Now observing changes on", u.String())
+
+	atomic.AddUint64(&(c.runID), 1)
+	currentRunID := atomic.LoadUint64(&(c.runID))
+
+	go func(runID uint64, u *url.URL) {
+		since := time.Now().Unix() * 1000
+		for {
+			pr, err := c.fetchEvents(since)
+
+			if err != nil {
+				fmt.Println(err)
+				fmt.Printf("Reattempting to connect to %s in %d seconds \n", u.String(), c.Reattempt)
```

Comment:
Should this use log instead of fmt?  And should we make logging enabled/disabled a bool setting on the client?  I'm not sure what's the idiomatic approach for this.  With the longpoll server, it uses log and has a bool to whether or not it logs.